core/trino-main/src/main/java/io/trino/sql/planner/Coercer.java

```diff
@@ -48,6 +50,8 @@ public static Expression addCoercions(Expression expression, Map<NodeRef<Express
 
         public Rewriter(Map<NodeRef<Expression>, Type> coercions, Set<NodeRef<Expression>> typeOnlyCoercions)
         {
+            // ExpressionAnalyzer checks for explicit Cast to unknown. The check here is to prevent engine from adding such invalid coercion.
+            checkArgument(coercions.values().stream().noneMatch(UNKNOWN::equals), "Cannot add coercion to UNKNOWN");
```

Comment:
I know it's convenient to add this here, but the check really belongs in the `addCoercions` methods in the enclosing class. It's illegal to call them with such a coercion, independently of how the Rewriter works internally.

Reply:
makes sense